### PR TITLE
Ollama API endpoint parallel agent Visualization

### PR DESCRIPTION
**Pull request by 6rz6**
Added the infrastructure including the v1/completions Ollama Endpoint (Open AI compatible, just replace the APIs endpoint) Multi threaded loading of the Models from a local or remote server. (In progress) Parallel Visualization of 4 active agents including display of resource consumption per each Ollama model in real time Use of Quantized, highly focused trained mixture of models to achieve many correct fast interactions thus improving the final desired result - i call it ("Decentralized Mixtral 8x7b") <- which is by far my favorite model - by mistral AI

### CODE REVIEW

**Comment by captivus** [comment]
Hello!  Please clarify the features you propose to add via this PR, along with their specific benefits to GPT-Engineer.  Also, please ensure that all of the checks are successful.

**Comment by 6rz6** [comment]
Hi, I'm adding optimized parallel Ollama model loading, with an optional UI to assess the load the the models on the system.
The goal is to improve the performance of multiply agent working in parallel each with the ability to use a task specific fine tuned quantized LLM.
This will allow us to assess the minimal system requirements to run in the future on 'smaller' devices and get results in an adequate time frame, while recommending the user based on his HW what are the Ollama models which will perform optimally on his device.
Feel free to brainstorm with me, and any feedback would be highly welcome.

**Comment by 6rz6** [comment]
running the exact same command on my localmachine and it passes all green:
image

I dont get it, need to understand what is being checked differently with github pre-commit

**Comment by 6rz6** [comment]
I still get no hook id, will investigate more
image

**Comment by 6rz6** [comment]
HI how are you?
I believe i found the reason why check 0.3 failed, it the alias of my py Vs the python ver of the alias of your py .
i run 3.10.13 when running any python and the one on your server was set to 3.8.
as you wrote in the branches the support for v.23 was the last for python 3.8 and i was running all this rime on 3.10.13 and failing a test which all it does is basically to check the support of the file in python, in this case it was tes3 and that drove me insane all last night.

i can run it locally and i can also call the module locallyand in both cases since it m interpreter everything woek fine. just we when tried to run the python test. we were border line because i assume you just recently upgraded to py3.10-12 and probably it slipped of your mind.

I managed to run it today both from your server and my local both went smooth. all i had to do to solve the issue is to change *inmy session ofc not in your bashrc, the aliass of your py to 3.10.11.

i tried to do more tests today but the module is already in main which i cannot acesss and also i lost access to the branch you opened for me... :)

talk to me later when available, we can sync a bit

**Review by 6rz6 — COMMENTED** [review]
** #The code works well, tested it today on multiple machines**

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# GPT-Engineer
+# Ollama-Multi-threaded-UI-Integration for GPT-Engineer
```
Comment (6rz6) [inline README.md:1]: changed from @6rz6 [Ollama_Models_Arena.py](https://github.com/gpt-engineer-org/gpt-engineer/pull/1098/files#diff-28d7c8481b3d39311b7059628a1d9020b6d8a6e95e87e0815ac24976d4c6cc31) 

Integrated on v.0.28->0.29

**Comment by 6rz6** [comment]
ok, passed and integrated

[Tox pytest all python versions](https://github.com/gpt-engineer-org/gpt-engineer/actions/workflows/ci.yaml)

File: Ollama_Models_Arena.py
Comment (6rz6) [inline Ollama_Models_Arena.py:1]: @6rz6  [6rz6:6rz6-Ollama-Multi-threaded-UI-Integration](https://github.com/gpt-engineer-org/gpt-engineer/pull/1098/files#diff-28d7c8481b3d39311b7059628a1d9020b6d8a6e95e87e0815ac24976d4c6cc31)
> 
> [Ollama_Models_Arena.py](https://github.com/gpt-engineer-org/gpt-engineer/pull/1098/files#diff-28d7c8481b3d39311b7059628a1d9020b6d8a6e95e87e0815ac24976d4c6cc31)
> 
> Integrated on v.0.28->0.29

File: README.md (hunk 1 above)
Comment (TheoMcCabe) [reply README.md:1]: Your changing the title of the readme?

**Comment by TheoMcCabe** [comment]
I dont understand the point of this code to be honest.  What does it have to do with GPT Engineer?

**Comment by ErikBjare** [comment]
I don't think this makes sense to merge either. We're trying to keep the repo as lean as possible, and this has very little connection with gpt-engineer itself.

If you want to make this a tool for others to use, consider making your own repo :)

**Pull request closed without merging**